src/iperf_udp.c

```diff
@@ -212,52 +279,111 @@ iperf_udp_recv(struct iperf_stream *sp)
 int
 iperf_udp_send(struct iperf_stream *sp)
 {
-    int r;
-    int       size = sp->settings->blksize;
+    int r = 0;
+    int size = sp->settings->blksize;
     struct iperf_time before;
+    char *buf = sp->buffer;
+    uint32_t  sec, usec;
 
-    iperf_time_now(&before);
+#ifdef HAVE_SENDMMSG
+    int i, j, k;
+    char *b;
+#endif /* HAVE_SENDMMSG */
+
+#ifdef HAVE_SENDMMSG
+    /* if sendmmsg is used - set buffer pointer to next buffer */
+    if (sp->settings->send_recvmmsg == 1) {
+        i = sp->sendmmsg_buffered_packets_count++;
+        sp->msg[i].msg_hdr.msg_iovlen = 1;
+        buf = sp->pbuf;
+        sp->msg[i].msg_hdr.msg_iov->iov_base = buf;
+        sp->pbuf += sp->settings->blksize;
+        sp->msg[i].msg_hdr.msg_iov->iov_len = size;
+    }
+#endif /* HAVE_SENDMMSG */
 
+    /* Set message packet count */
     ++sp->packet_count;
-
     if (sp->test->udp_counters_64bit) {
-
-	uint32_t  sec, usec;
 	uint64_t  pcount;
-
-	sec = htonl(before.secs);
-	usec = htonl(before.usecs);
 	pcount = htobe64(sp->packet_count);
-
-	memcpy(sp->buffer, &sec, sizeof(sec));
-	memcpy(sp->buffer+4, &usec, sizeof(usec));
-	memcpy(sp->buffer+8, &pcount, sizeof(pcount));
-
+	memcpy(buf+8, &pcount, sizeof(pcount));
+    } else {
+	uint32_t  pcount;
+	pcount = htonl(sp->packet_count);
+	memcpy(buf+8, &pcount, sizeof(pcount));	
     }
-    else {
-
-	uint32_t  sec, usec, pcount;
 
-	sec = htonl(before.secs);
-	usec = htonl(before.usecs);
-	pcount = htonl(sp->packet_count);
+/* Use sendmmsg when approriate to send the packet, else use Nwrite */
+#ifdef HAVE_SENDMMSG
+    if (sp->settings->send_recvmmsg == 1) {
+	/* When enough "burst" packets were buffered - send them */
+	if (sp->sendmmsg_buffered_packets_count >= sp->settings->burst) {
+            /* Set actual sending time to all packets*/
+            iperf_time_now(&before);
+            sec = htonl(before.secs);
+	    usec = htonl(before.usecs);
+            for (i = 0; i < sp->sendmmsg_buffered_packets_count; i++) {
+                b = sp->msg[i].msg_hdr.msg_iov->iov_base;
+                memcpy(b, &sec, sizeof(sec));
+		memcpy(b+4, &usec, sizeof(usec));	
+	    }
 
-	memcpy(sp->buffer, &sec, sizeof(sec));
-	memcpy(sp->buffer+4, &usec, sizeof(usec));
-	memcpy(sp->buffer+8, &pcount, sizeof(pcount));
+	    /* Sending messages and making sure all packets are sent */
+	    i = 0;	/* count of messages sent */
+	    r = 0;	/* total bytes sent */
+	    while (i < sp->sendmmsg_buffered_packets_count) {
+	        j = sendmmsg(sp->socket, &sp->msg[i], sp->sendmmsg_buffered_packets_count - i, MSG_DONTWAIT);
+		if (j < 0) {
+		    r = j;
+		    break;
+		}
```

Comment:
Did you forget to `i += j;`?

Reply:
This is done below in line 349.  The operation is delayed since the original `i` value is required in the `while` body.